### ISSUE TITLE
Close widgets before tearing down other fixtures

### ISSUE THREAD

**Issue by The-Compiler** [pytest-dev/pytest-qt#106]
In https://github.com/The-Compiler/qutebrowser/pull/1048#issuecomment-150660769 I noticed a problem: I use a fixture to provide a `config` object in my object registry (basically a dict mapping strings to global objects).

However, on OS X, closing the widget seems to call `minimumSizeHint` on it, which causes the widget to get the config - but at the point `qtbot` calls `close` on the widget, the other fixture already is torn down.

I think the widget cleanup from `qtbot` should happen as early as possible.

**Comment by nicoddemus** [comment]
Agree! I will try to work on this along the week. :smile:

This seems a duplicate of #105, no?

**Comment by The-Compiler** [comment]
I don't think they're duplicates - their cause is similar, but they're two different things, and showing up very differently.

This issue means `QWidget.close` can be called on a widget after e.g. a monkeypatching or mocking is already undone, which can cause all kind of funny effects depending on what the widget does in `close`/`closeEvent`.

#105 is about exceptions not being captured if they happen during setup/teardown, which will probably cause it to be ignored with PySide/PyQt4, and causes an `abort()` without any information (because of stdout capturing) on PyQt5.

**Comment by nicoddemus** [comment]
You're right, thanks for the clarification.

**Comment by The-Compiler** [comment]
I don't think #108 fixes this - here is a reproducer which still fails with #108:

```python
import pytest
from PyQt5.QtWidgets import QWidget


class Widget(QWidget):

    def __init__(self, parent=None):
        super().__init__(parent)
        self.attr = False

    def closeEvent(self, e):
        assert self.attr
        e.accept()


@pytest.yield_fixture
def widget(qtbot):
    w = Widget()
    qtbot.add_widget(w)
    w.attr = True
    yield w
    w.attr = False


def test_foo(widget):
    pass
```

Of course this is just a contrived example - but in my real code, I had issues
on OS X because this happens:

- I have a widget which needs a configuration value in `sizeHint`.
- I have a `yield_fixture` which provides a fake configuration and cleans it up after the `yield`
- The test ends, the fixture is torn down
- Now `qtbot` calls `close` on the widget (rather than *before* the last step)
- `close` calls `sizeHint` internally (on OS X) for some reason
- The configuration isn't there anymore -> boom :wink:

#105 is about this exception causing `abort()` rather than being captured - that part is fixed with #108.

This issue is about `Widget.close()` being called after the `widget` fixture is torn down, which still happens with #108.

I hope this clears things up :smile:

**Comment by nicoddemus** [comment]
Oh now I get. Sorry! :smile:

Seems like all it would require is to close the widgets at `pytest_runtest_teardown`, right after a `process_events` but before the `yield`. I will work on this later.
